### PR TITLE
feat: BasicoModel Instantiation & Use Relative Imports

### PR DESCRIPTION
**Pull request by Rakesh-Seenu**
# For authors

This PR addresses key improvements based on feedback from the previous PR. The main changes include:

- Using relative imports instead of absolute paths for better maintainability and modularity.
- Ensuring BascoModel follows a singleton pattern, meaning only one instance of the model is created per unique model name. This prevents redundant object creation and optimizes resource usage.
- Code cleanup to align with previous review comments, ensuring consistency and best practices.

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [ ] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [ ] My code follows the style guidelines mentioned in the Code/DevOps guides
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (e.g. MkDocs)
- [ ] My changes generate no new warnings
- [ ] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

**Comment by Rakesh-Seenu** [comment]
@gurdeep330 Please let me know if any changes are required.

File: app/data/markdown_report.py [hunk 1]
```diff
@@ -31,14 +31,16 @@ def generate_markdown_report(input_file: str,
...
         output_file (str): Path where the markdown report will be saved.
     """
-    biomodel = BasicoModel(sbml_file_path=input_file)
-    metadata_list = [
-        biomodel.get_model_metadata(file)
-        for file in os.listdir(input_file)
-        if file.endswith(".xml")
-    ]
+    metadata_list = []
+    for file in os.listdir(input_file):
```
Comment (gurdeep330) [inline app/data/markdown_report.py:35]: Shouldn't it be called input folder?

File: app/data/markdown_report.py [hunk 2]
```diff
@@ -31,14 +31,16 @@ def generate_markdown_report(input_file: str,
         template_file (str): Template file name.
         output_file (str): Path where the markdown report will be saved.
```
Comment (gurdeep330) [inline app/data/markdown_report.py:32]: Shouldn't it be called output_folder?

File: app/leaderboard/table.py [hunk 3]
```diff
@@ -6,23 +6,21 @@
...
     """Main function using Hydra for configuration management."""
     # Load hydra configuration
     with hydra.initialize(version_base=None, config_path="../configs"):
-        cfg = hydra.compose(config_name="config", 
-                            overrides=["leaderboard=default"]) 
-    
+        cfg = hydra.compose(config_name="config",
+                            overrides=["leaderboard=default"])
     cfg = cfg.leaderboard
     input_file = cfg.paths.input_file
```
Comment (gurdeep330) [inline app/leaderboard/table.py:18]: Same as above

File: app/leaderboard/table.py [hunk 4]
```diff
@@ -6,23 +6,21 @@
...
     template_dir = cfg.paths.template_dir
     template_file = cfg.paths.template_file
     output_file = cfg.paths.output_file
```
Comment (gurdeep330) [inline app/leaderboard/table.py:21]: Same as above

File: tests/app/data/test_markdown.py [hunk 5]
```diff
@@ -16,6 +16,19 @@ def test_validate_sbml_folder_path_success():
     model = BasicoModel(sbml_file_path="vpeleaderboard/data/models")
```
Comment (gurdeep330) [inline tests/app/data/test_markdown.py:16]: Isn't this supposed to take the path to the file and not just the folder?

File: tests/app/data/test_markdown.py [hunk 6]
```diff
@@ -16,6 +16,19 @@ def test_validate_sbml_folder_path_success():
...
+    Test that a ValueError is raised when the sbml_folder_path does not exist.
+    """
+    non_existent_folder = "vpeleaderboard/data/non_existent_models"
+    # Ensure the folder does not exist
+    assert not os.path.exists(non_existent_folder), \
+        f"Test setup issue: {non_existent_folder} should not exist."
+
+    # Expect ValueError due to missing folder
+    with pytest.raises(ValueError, match=f"SBML file not found: {non_existent_folder}"):
+        BasicoModel(sbml_file_path=non_existent_folder)
```
Comment (gurdeep330) [inline tests/app/data/test_markdown.py:30]: folder or file?

File: tests/utils/utils.py [hunk 7]
```diff
@@ -11,21 +11,14 @@
 @pytest.fixture
 def model_fixture():
     """Fixture for the BasicoModel class."""
-    return BasicoModel(sbml_folder_path="vpeleaderboard/data/models")
+    return BasicoModel("vpeleaderboard/data/models")
```
Comment (gurdeep330) [inline tests/utils/utils.py:14]: folder or file?

File: tests/vpeleaderboard/data/test_basico_model.py [hunk 8]
```diff
@@ -15,22 +15,12 @@ def test_validate_sbml_folder_path_success():
     model = BasicoModel(sbml_file_path="vpeleaderboard/data/models")
```
Comment (gurdeep330) [inline tests/vpeleaderboard/data/test_basico_model.py:15]: folder or file?

File: vpeleaderboard/data/src/basico_model.py [hunk 9]
```diff
@@ -20,52 +17,44 @@
...
+    Ensures a single instance per component.
     """
-    sbml_file_path: Optional[str] = Field(None, description="Path to an SBML files folder")
-    simulation_results: Optional[pd.DataFrame] = None
-    name: Optional[str] = Field("", description="Name of the model")
-    description: Optional[str] = Field("", description="Description of the model")
-
-    copasi_model: Optional[object] = None
-    model_config = {"arbitrary_types_allowed": True}
+    sbml_file_path: Optional[str] = Field(None, description="Path to an SBML file ")
```
Comment (gurdeep330) [inline vpeleaderboard/data/src/basico_model.py:22]: Should it still be optional?

**Review by gurdeep330 — CHANGES_REQUESTED** [review]
Thanks for fixing the issues, @Rakesh-Seenu 

My only major concern is that I sometimes see the BasicoModel object takes the path to folder as input. I suggest this be replaced with the path to a file.

File: vpeleaderboard/data/src/test/test.ipynb
Comment (gurdeep330) [inline vpeleaderboard/data/src/test/test.ipynb:1]: Is this file needed?

**Review by gurdeep330 — CHANGES_REQUESTED** [review]

File: vpeleaderboard/data/src/test/test.ipynb
Comment (Rakesh-Seenu) [reply vpeleaderboard/data/src/test/test.ipynb:1]: no, we don't need this file

**Review by gurdeep330 — APPROVED** [review]
